### ISSUE TITLE
Extend triggers with new time based events

### ISSUE THREAD

**Issue by fliebe92** [zammad/zammad#4484]
### Actual behavior:

Currently, in the trigger configuration, it is not possible to select e.g. when a ticket reminder is reached or when a ticket is going to escalate (triggers are also not fired for these events).

### Expected behavior:

The trigger configuration is going to be extended to support time-based events.

A new dropdown field will be added to triggers called **Activated by**, with _Action_ and _Time event_ options.

<img width="958" alt="Screenshot 2023-04-03 at 11 28 00" src="https://user-images.githubusercontent.com/6049445/229484508-3faeae54-0214-471f-95b6-230a6b3f103c.png">
<img width="958" alt="Screenshot 2023-04-03 at 11 28 18" src="https://user-images.githubusercontent.com/6049445/229484516-e88687b2-6955-4fa8-acc9-5c4f30caa6d1.png">
<img width="221" alt="Screenshot 2023-04-03 at 11 28 08" src="https://user-images.githubusercontent.com/6049445/229484537-dd55f1e2-af82-4d9b-9362-e0164c905841.png">

If the value is _Action_, a new field below it can be used to select the action execution behavior (**Action execution**):

* _Selective_ (default behavior, after #4550 is merged)
* _Always_ (execute whenever conditions are matching)

<img width="955" alt="Screenshot 2023-04-03 at 11 28 30" src="https://user-images.githubusercontent.com/6049445/229484710-8b16f246-a76c-401a-aebb-2bbe428e7f08.png">

If the value is _Time event_, the following attributes in the conditions will receive new possible values:

* `Pending till` is extended by values `has reached`
* `Escalation at` is extended by values `has reached` and `has reached warning`

<img width="953" alt="Screenshot 2023-04-04 at 10 20 11" src="https://user-images.githubusercontent.com/6049445/229747193-2dbdd1f1-8948-4de9-909f-6a8af07734eb.png">

Additionally, `Action` condition will not be available anynmore.

By introducing this new behavior, we will not be breaking existing trigger configurations.

### Technical Implementation/Questions:

- The new _Time event_ triggers will be executed when a pending reminder is reached or a ticket is escalated (or escalation warning time is reached).
- The triggers will have to be restricted in order not to run every 15/5 Minutes (which is the default schedule time for pending/escalation jobs).

```[tasklist]
### Tasks
- [x] Implement another `Transaction::Backend::Async` module for new triggers (@mantas)
- [x] Add support for new time event triggers in `Ticket#perform_triggers` and filter only for relevant triggers (@mantas)
- [x] Modify `Ticket#selectors` to support new operators based on the passed parameters (@mantas)
- [x] Re-use ticket history to remember already executed triggers for a ticket + field combination (@mantas)
- [x] Add new trigger attribute for context (backend + GUI) (@mantas)
- [x] Rename context attribute to activator and align GUI with the mockup (@dvuckovic)
- [x] Add new attribute for execution condition mode (backend + GUI) (@dvuckovic)
- [x] Switch to hiding of the complete execution condition mode control in case of time event triggers (@dvuckovic)
- [x] Add new operator types to affected fields in the ticket selector GUI (@dvuckovic)
- [x] Extend tests (RSpec trigger time events + QUnit ticket selector + Capybara manage triggers) (@dvuckovic)
- [x] Extend tests (Trigger new methods) (@mantas)
- [x] Extend tests (Ticket history) (@mantas)
- [x] Cleanup coffeescript implementation (@mantas)
- [x] Connect execution condition mode attribute with expected behavior behind required changes in ticket selector (@mantas, @dvuckovic)
- [x] Review (@rolfschmidt)
```

**Comment by YetAnotherGerrit** [comment]
I mostly use schedulers for scenarios like yours. Running every 10 minutes and checking if a ticket has escalated/been reminded within the last 10 minutes.

But having those events also trigger "triggers" would be convenient. But two things to consider:
- adding those events to possible actions could result in unexpected behaviour in existing systems. May be a breaking change, may it?
- having then 7 instead of 4 actions, it would be nice to have this as a multi-punch:

![grafik](https://user-images.githubusercontent.com/97293098/216610779-1bfb775c-3c48-44a3-86ee-6bb305e7e913.png)

instead of

![grafik](https://user-images.githubusercontent.com/97293098/216610141-0f404fac-876f-4a94-b187-e3bc8f242144.png)

**Comment by MrGeneration** [comment]
That is a possible breaking change. Right now ticket updates only count as such if they where caused by an user.
https://admin-docs.zammad.org/en/latest/manage/trigger/limitations.html

While this enables users to a lot more options, this may break existing setups quite badly by getting super noisy and thus causing a lot of support overhead.

**Comment by fliebe92** [comment]
@YetAnotherGerrit @MrGeneration thanks for your input, we are going to do another round of refinement for this enhancement to not break existing setups. Stay tuned! 😎

**Comment by mantas** [comment]
There's an edge case with AND/OR conditions. For example:

```
A condition -> title is "QWERTY" AND pending has reached
OR
B condition -> title is "ABC"
```

When ticket reaches pending time, trigger shall be executed for A condition only. But it's difficult to detect that ticket matches specifically A condition and not B.

Unfortunately there seems to be no easy way to detect exact matching condition using current SQL WHERE clause approach.

Alternatively we could use static analysis of the query and try one of the following approaches:

a) disallow usage of `has reached` in OR queries
b) pre-format condition and remove OR parts where `has reached` is not mentioned

Both seems to be prone to breaking though.

Should we put in effort to ensure that such edge cases work correctly?

**Comment by mantas** [comment]
... however, after rubber-ducking in above comment, I'm thinking the `b)` option to pre-modify condition might work...!

**Comment by mgruner** [comment]
Can you explain what kind of SQL condition would be needed?

**Comment by mantas** [comment]
@mgruner found one more issue. Pending time and escalation warnings are processed every 15 minutes and then `History` item is used to send them once a day at most.

We need a similar mechanism to limit executing of `has reached` triggers then. Would they be executed once-a-day in a similar way? Or use a different timing?

**Comment by dominikklein** [comment]
Yes, this topic was already mentioned in the initial description and one of the things which was open for clarification.

**Comment by mantas** [comment]
Yes. But it's not exactly clear what should be the outcome. E.g. mimic notifications to perform once a day via history entry. Or something else.

**Comment by mantas** [comment]
Notes of 2023-03-31  refinement call:

- Triggers will be explicitly action-based (= existing functionality) or time-based (= new functionality).

- Action-based triggers will work same way as before.

- Time-based triggers will be executed on pending time and escalation + escalation warning. Later will be based on `escalation_at` time only, no `first_response/close/update` granular escalations.

- Time-based triggers  will be executed on all 3 events (pending time, escalation and escalation warning). Trigger condition will have `has reached` selector which will allow to limit to limit to a selected event.

- Triggers will fire no more than once a day for a given ticket in each event. A new DB table will be used to keep track of which trigger was already fired.

- Each event (pending time, escalation and escalation warning) will be treated separately. So technically a trigger may fire 3 times in day for a ticket, once for each event.

- In time-based triggers, it won't be possible to select `Action` condition. Since there's no associated create/update/merge action

**Comment by dominikklein** [comment]
> Triggers will fire no more than once a day for a given ticket in each event. A new DB table will be used to keep track of which trigger was already fired.

Maybe a new table is not needed, because we have this enhancement idea about the history, when this will really be implemented, we have maybe duplicated data with this new table.

https://github.com/zammad/coordination-private-features/issues/16

Only as a side note, I'm not sure what the best solution is now.

**Comment by mgruner** [comment]
We could check if it's feasible just to add support for Triggers in History at this point, not the others. @mantas @dvuckovic would be interesting to compare the ballparks.

**Comment by mantas** [comment]
Could be possible. Not the cleanest way, but won't require too hacky approach either.

Then fired triggers would show up in ticket history view, right?

**Comment by fliebe92** [comment]
> Could be possible. Not the cleanest way, but won't require too hacky approach either.
> 
> Then fired triggers would show up in ticket history view, right?

Exactly, but this is - at least for me - "nice-to-know" information and could help agents to understand what happened.

**Comment by YetAnotherGerrit** [comment]
Could this be an opportunity to re-arrange the existing mess with macros/triggers/schedulers as well? All three are very similar from a user perspective, it's just the activator that differs. In Power Automate all would be called "flows", "automated flows" (trigger), "instant flows" (macros) and "scheduled flows".

https://learn.microsoft.com/en-us/power-automate/flow-types#cloud-flows

I might be imprinted with that distinction, because prior to Zammad, I worked a lot with Power Automate, but it is very intuitive and makes much more sense than our approach to list them completely separate in our UI.

I wouldn't copy the term "flows", but it makes sense to bring them closer together in the UI and in terminology.

**Comment by dvuckovic** [comment]
Task list in the OP was updated with the new approach. @YetAnotherGerrit was instructed to move the excellent "unification" idea to a separate enhancement process, since it's out-of-scope for this issue.

**Comment by dvuckovic** [comment]
Follow-up for changing the behavior of one-per-day notifications will be handled in https://github.com/zammad/zammad/issues/4348

The new behavior will be aligned and cover the time-based triggers as well, for consistency purposes.

**Comment by dvuckovic** [comment]
@KoJie @martini A proposal for rewording of option in the trigger action execution mode selection: adding `(default)` suffix to the _Selective_ option.

![Screen Shot 2023-04-24 at 09 30 58](https://user-images.githubusercontent.com/6049445/234095980-09c33b3e-044d-4d98-bf7d-c00e9b74c24a.png)

What do you think?
